### PR TITLE
Fix keyboard shortcuts

### PR DESCRIPTION
**Pull request by victorporof**
tl;dr: We can't handle some keyboard shortcuts in frontends. We need some amount of platform support to overwrite OS-level shortcuts like Cmd+W on macOS. And down the rabbit hole we go.

Problem statement:
We'd like to have some keyboard shortcuts. Registering them in the frontend is preferable simply due to convenience. It's easy to listen to key events and do things in the event listeners.

Solutions:
1. A naive approach would register all keyboard shortcuts in the frontend alone. This cannot work in all cases though, since we can't alter platform-specific behavior (like Cmd+W hiding windows on macOS, while we want it to close tabs instead; maybe we'll even decide to do something different on Cmd+Q or Alt+F4, who knows).

2. Another approach would be (1), plus a simple fix to the problem described above, where windows are prevented from closing unless specifically directed by the frontend. On the surface, this works fine, since we draw our own window controls (close button etc.) and can instruct the browser runner platform to quit from the frontend itself. This cannot scale, however, since the frontend isn't the only responsible entity for closing the window. (Ctrl+C in a terminal should work, we can't allow that to happen while still preventing windows to close generally; Additionally, maybe we want to draw native window controls on some operating systems). There aren't any good solutions to handling this problem, and once you add other behaviors in the mix (suppose we'd like to override Cmd+M for custom actions instead of minimizing windows, for example, similarly to how Cmd+W doesn't hide windows, but performs the custom action of closing a tab), possible implementations (if they even exist) rapidly increase in complexity. This was the solution implemented in #51 

3. Yet another approach would be (1), plus a subtly different take on (2): instead of preventing windows from closing, eat the overridden keyboard shortcuts themselves on the platform-side. In theory this would allow custom behavior on Cmd+W, for example, when defined in the frontend, while the platform would prevent the default OS-level action of hiding a window. In practice this doesn't work, as frontends would not receive those key combos at all if they're eaten at the platform level.

4. Accept the reality that overriding os-defined keyboard shortcuts requires platform support, and adjust accordingly. 

This PR implements (4). Generally it's expected that most keyboard shortcuts can be implemented by simply listening to key events in the frontend. However, in some cases we simply cannot rely on this to work cross operating systems (e.g. Cmd+W), so provide a means of registering shortcuts through browser runners, but still handle them in the frontend. Therefore the proposed solution is a mixed approach.

A possible emerging advantage from (4) is that we're not forced to perform the actions we want (based on what keyboard shortcuts are pressed) in the same process in which they're registered, so we're not constrained nor concerned with having to sync or share application state between components/processes. Registering keyboard shortcuts is a notification-like service provided by browser runners, which you can use if you want to.

I'd also be happy if we're to decide we want all shortcuts to be handled through the browser runner completely, not just Cmd+W and the likes. However, this necessarily requires that all browser runner implementations accept keyboard shortcut registering commands, which makes it a bit developer unfriendly to get one such implementation quickly running off the ground. For example, if we're to go this route, our qbrt runner will never have any keyboard shortcuts unless we explicitly add support for it. With the mixed approach suggested in this PR, Cmd+W hides windows instead of closing tabs in qbrt, but at least everything else does work.

### CODE REVIEW

File: src/browser-frontend/sagas/server-event-handlers.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+*/
+
+import { takeEvery, call, put, select } from 'redux-saga/effects';
+
+import SharedActions from '../../shared/actions/shared-actions';
+import PagesModelActions from '../actions/pages-model-actions';
+import * as DomainPagesSelectors from '../selectors/domain-pages-selectors';
+import * as UIPagesSelectors from '../selectors/ui-pages-selectors';
+
+function* handleCmdOrControlQ({ meta: client }) {
```
Comment (victorporof) [inline src/browser-frontend/sagas/server-event-handlers.js]: Note: handling this shortcut isn't strictly necessary, as it does the same thing the OS would do. We can remove this if desired, it's here just to mirror the previous implementation (and prove a point I suppose).

Comment (victorporof) [reply src/browser-frontend/sagas/server-event-handlers.js]: Note: there's a fair bit of plumbing going on for this to happen. The request to register keyboard shortcuts is sent from the browser frontend to the browser server, which pipes it onward to the browser runner. When keys are pressed, the event bubbles back the same route: runner -> server -> frontend. This implementation is in line with the proposed architecture in #31, following a "single two-lane street" principle where all communication between all components follows a single path/bus, instead of allowing all components to communicate with each other as they would like. In principle, this should reduce complexity in the long run.

File: src/browser-frontend/views/browser/window.jsx [hunk 2]
```diff
@@ -17,9 +17,8 @@ import Mousetrap from 'mousetrap';
 
 import { client } from '../../global';
 import PagesModelActions from '../../actions/pages-model-actions';
-// import * as SharedPropTypes from '../../model/shared-prop-types';
-// import SharedActions from '../../../shared/actions/shared-actions';
-// import * as UIPagesSelectors from '../../selectors/ui-pages-selectors';
```
Comment (victorporof) [inline src/browser-frontend/views/browser/window.jsx]: Commented out parts in this PR are part of the backouts performed in f148577c0ed17778d4f7432dfca4c4094a1d8ed7 and d5bbd159274ad0ab54243fc84bdb90204219e942

File: src/browser-runner/platform-electron/sagas/server-command-handlers.js [hunk 3]
```diff
@@ -86,6 +79,22 @@ function* openDevTools({ meta: client, payload: { winId, detach } }) {
   yield call([client, client.send], SharedActions.events.fromRunner.toServer.app.window.devtools.closed({ winId }));
 }
 
+function* registerKeyShortcuts({ meta: client, payload: { winId, shortcuts } }) {
```
Comment (victorporof) [inline src/browser-runner/platform-electron/sagas/server-command-handlers.js:82]: This is where we register shortcuts in electron-runner. These commands/actions originate in the browser frontend.

File: src/browser-frontend/sagas/server-event-handlers.js (hunk 1 above)
Comment (bwinton) [reply src/browser-frontend/sagas/server-event-handlers.js]: Just as a note, we could follow Firefox's conventions, and call it `handleAccelQ`…  😉
(Actually, we can't because of electron-localshortcut.  But I would prefer the shorter `CmdOrCtrl+Q`.)

Also, your long note should probably be a comment somewhere.  Either in the code, or in the architecture document would work for me.

File: src/browser-frontend/sagas/server-event-handlers.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    client.send(SharedActions.events.fromFrontend.toServer.app.window.requestedClose());
+    return;
+  }
+
+  const selectedPageId = yield select(UIPagesSelectors.getSelectedPageId);
+  yield put(PagesModelActions.removePage({ pageId: selectedPageId }));
+}
+
+function* onKeyShortcutPressed({ meta: client, payload: { shortcut } }) {
+  if (shortcut.keys === 'CommandOrControl+Q') {
```
Comment (bwinton) [inline src/browser-frontend/sagas/server-event-handlers.js]: In the distant past, I had a lot of problems with long chains of `if/else if` blocks.  Maybe this one won't grow too much further, but if we add another if, I would suggest switching it to a map of strings to functions…

File: src/browser-frontend/views/browser/window.jsx [hunk 5]
```diff
@@ -33,19 +32,20 @@ import Content from './content';
 })
 export default class Window extends PureComponent {
   componentDidMount() {
+    // Some keyboard shortcuts need to be registered at the platform level,
+    // because they clash with OS-level shortcuts (e.g. Cmd+W on macOS
+    // hides windows, but we want it to close tabs instead.). The platform
+    // is tasked with overriding these shortcuts and providing us control.
```
Comment (bwinton) [inline src/browser-frontend/views/browser/window.jsx]: I do wonder a little about having two different places to configure key events.  Would it be terrible to put `CommandOrControl+T` and the Konami code in the platform, too?  Or move them to a saga?

**Review by bwinton — APPROVED** [review]
Apart from the small stuff I mentioned, this looks good to me.
(Oh, and I tested it on Windows, and it works great, too! 🙂)

File: src/browser-frontend/views/browser/window.jsx (hunk 5 above)
Comment (victorporof) [reply src/browser-frontend/views/browser/window.jsx]: It would not be terrible at all! As described in the top comment, I'd also be happy if we're to decide we want all shortcuts to be handled through the platform completely, not just some of them (Cmd+W for example). However, that introduces some developer effort with making all browser runners (including qbrt) support keyboard shortcut notifications. We want to do this anyway in the long run, though.

File: src/browser-frontend/sagas/server-event-handlers.js (hunk 4 above)
Comment (victorporof) [reply src/browser-frontend/sagas/server-event-handlers.js]: Eh, I don't mind either way.

File: src/browser-frontend/sagas/server-event-handlers.js (hunk 1 above)
Comment (victorporof) [reply src/browser-frontend/sagas/server-event-handlers.js]: Sounds good! Will rename.

Comment (victorporof) [reply src/browser-frontend/sagas/server-event-handlers.js]: Regarding documenting this, a lot more documentation is coming soon, and this will be included.

**Comment by victorporof** [comment]
Addressed @bwinton's comments.

**Comment by victorporof** [comment]
@bgrins I've addressed your comments from our conversation yesterday. The changes I've made are in the "Post: ..." commits. We can now identify websocket connections to the frontend by window ids, so if we decide to move "The List of Nasty Keyboard Shortcuts" in the runners, and emit them without them being requested, we could. Also, I've moved all keyboard shortcut handlers into sagas, and as such we're now listening to all key events normally (including the "nasty ones"), while also calling `preventDefault()`.

Hopefully this addresses your concerns. The only thing remaining is deciding whether or not "nasty shortcuts" are to be subscribed to explicitly, or emitted by default by all browser runners. My hope is that we can move this discussion in a followup.

File: src/browser-frontend/sagas/server-event-handlers.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+import { takeEvery, put } from 'redux-saga/effects';
+
+import SharedActions from '../../shared/actions/shared-actions';
+import KeyboardShortcutsActions from '../actions/keyboard-shortcuts-actions';
+
+function* onKeyShortcutPressed({ payload: { shortcut } }) {
```
Comment (bgrins) [inline src/browser-frontend/sagas/server-event-handlers.js]: This could use a comment explaining why these are coming from the server

File: src/browser-server/sagas/frontend-command-handlers.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+function* registerKeyShortcuts({ meta: frontendConn, payload: shortcuts }) {
+  const frontendConnId = frontendConn.id;
+  const { runnerConnId, winId } = yield select(FrontendConnectionsSelectors.getFrontendClientMetaData, frontendConnId);
+
+  const action = SharedActions.commands.fromServer.toRunner.app.window.keyShortcuts.register({
+    winId,
+    shortcuts,
+  });
+
+  const runnerConn = Connection.getWithId(runnerConnId);
```
Comment (bgrins) [inline src/browser-server/sagas/frontend-command-handlers.js:28]: I'd expect we would be using `FrontendConnection.getWithFrontendWinId` here instead of `Connection.getWithId` (ditto for all the other places throughout browser-server that take in a frontendConn).  Or, if there's not additional value being added by handling these connections as FrontendConnections then maybe we don't need the new class at all and instead can make the `frontendWinId` an optional property on the plain `Connection` object (and add the `static getWithFrontendWinId` there as well) then just always deal with connections.

**Review by bgrins — APPROVED** [review]

File: src/browser-frontend/sagas/server-event-handlers.js (hunk 6 above)
Comment (victorporof) [reply src/browser-frontend/sagas/server-event-handlers.js]: I'll add one!

**Comment by victorporof** [comment]
@bgrins 

> I'd expect we would be using FrontendConnection.getWithFrontendWinId here instead of Connection.getWithId (ditto for all the other places throughout browser-server that take in a frontendConn). Or, if there's not additional value being added by handling these connections as FrontendConnections then maybe we don't need the new class at all and instead can make the frontendWinId an optional property on the plain Connection object (and add the static getWithFrontendWinId there as well) then just always deal with connections.

This wouldn't be correct. To sum things up, there are two kinds of connections we're dealing with: 
1. connections from the server to the frontend, and 
2. connections from the server to the runner. 

Every connection from the server to the frontend *can* be identified by a window id, because every frontend has their own websocket pipe open, and one frontend === one window. 

However, every connection from the server to the runner *cannot* be identified by a window id, because every runner has a single websocket pipe open, but can have display multiple frontend windows.

**Comment by ncalexan** [comment]
Sorry, late to this party, but I believe option 4) -- platform support -- is the only viable option for a full-fledged browser product.  We're changing platform behaviour; we need platform support.

I didn't review the code but I think the approach is sensible.  Onward!
